File: 92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        int i = 1;
        ListNode *p = head, *pre = nullptr, new_head(0), *t, *q;
        while (p && i < m) {
            pre = p;
            p = p->next;
            i++;
        }
            
        if (pre){ 
            q = pre->next;
        } else {
            q = head;
        }
        
        ListNode* tail = q;
        while (q && i <= n) {
            t = q;
            q = q->next;
            i++;
            t->next = new_head.next;
            new_head.next = t;
        }
        if (pre) {
            pre->next = new_head.next;
            tail->next = q;
            return head;
        } 
        
        tail->next = q;
        return new_head.next;

    }
};

File: python/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        st = []
        dummy = ListNode(0)
        p = dummy
        p.next = head
        idx = 1
        while head and idx!=m:
            p = head
            head = head.next
            idx += 1

        tmp = head
        while tmp and idx!=n+1:
            st.append(tmp)
            tmp = tmp.next
            idx += 1
            
        while st:
            p.next = st.pop()
            p = p.next
        
        p.next = tmp
        return dummy.next
        

        
if __name__ == '__main__':
    head = ListNode(3)
    head.next = ListNode(5)
    su = Solution()
    su.reverseBetween(head, 1, 2)